Update serial_core.c to create an anonymous parent device for NI-987x

nikhalim opened the pull request:
niserial987xDriver doesn't pass in a parent device pointer to uartPort when calling uart_add_one_port. This, in turn, crashes the kernel when serial_core attempts to locate the parent device pointer (since the pointer is NULL).

This PR attempts to use platform_device_register_simple to create a platform device that could be assigned as the parent device for NI-987x ports. It does this by checking if the caller is niserial987xDriver. If it is, it gets the slot number of the 987x module, and creates an appropriate platform_device for each uart port that needs to be registered.

nikhalim committed:
Update serial_core.c to create an anonymous parent device for NI-987x

niserial987xDriver doesn't pass in a parent device pointer to uartPort when calling uart_add_one_port. This, in turn, crashes the kernel when serial_core attempts to locate the parent device pointer (since the pointer is NULL).

This PR attempts to use platform_device_register_simple to create a platform device that could be assigned as the parent device for NI-987x ports. It does this by checking if the caller is niserial987xDriver. If it is, it gets the slot number of the 987x module, and creates an appropriate platform_device for each uart port that needs to be registered.

sjasonsmith commented on drivers/tty/serial/serial_core.c:
```diff
@@ -3404,6 +3404,38 @@ int serial_core_register_port(struct uart_driver *drv, struct uart_port *port)
+		if (strncmp(driver_name, "niserial987x", 12) == 0) {
+			// Max number of cRIO modules = 8
+			// Max number of uart ports per module = 4
+			const char *const platform_device_names[32] = {
```
If you make this a 2-D array then it will be a lot easier to index into using the form:
`platform_device_names[slot_number - 1][port->line]`

Probably make it static just to make sure this doesn't get initialized at runtime.
You could also use a shorter name just to waste less space.

```
static const char * const [8][4] = {
        { "987x_1_1", "987x_1_2", "987x_1_3", "987x_1_4" },
        ...
```

sjasonsmith replied:
Looking at how platform_device_register_simple works, it has to point to strings that will remain valid after the function returns, so it should definitely be static.

It would make sense to move it out of the function...but it is nice that in the current location it keeps the patch at one added block, and not multiple that have to merge in different places.

nikhalim replied:
Updated to 2D array

sjasonsmith commented on drivers/tty/serial/serial_core.c:
```diff
@@ -3404,6 +3404,38 @@ int serial_core_register_port(struct uart_driver *drv, struct uart_port *port)
+				goto err_unlock;
+			}		
+
+			port->dev = &anonymous_parent_dev->dev;
```
I think you need to remove this device in `serial_core_unregister_port`, which should be called when you remove a module. You will need to make sure it was created for a 987x device if 

I think right now it would try creating a device with the same name again.

nikhalim replied:
Fixed. Verified by inserting and removing modules multiple times.

sjasonsmith commented on drivers/tty/serial/serial_core.c:
```diff
@@ -3404,6 +3404,38 @@ int serial_core_register_port(struct uart_driver *drv, struct uart_port *port)
+			};
+			static struct platform_device *anonymous_parent_dev = NULL;
+			int slot_number = 0;
+			sscanf(driver_name, "niserial987xSlot%uDriver", &slot_number);				
```
sscanf returns the number of matched items. You should check to make sure it is 1, and error out if it is int.

sjasonsmith commented on drivers/tty/serial/serial_core.c:
```diff
@@ -3404,6 +3404,38 @@ int serial_core_register_port(struct uart_driver *drv, struct uart_port *port)
+			port->dev = &anonymous_parent_dev->dev;
+		}
+		
+		goto err_unlock;
```
I don't think this should go to `err_unlock`. This is the case where you successfully created a new device. You should be able to execute the rest of the function since you have a valid device now.

nikhalim replied:
Oops, I meant it to go into the else bracket.

nikhalim committed:
Update serial_core.c

Delete platform_device created for niserial987x in port is unregistered
Addressed PR feedbacks
